ouroboros-network/src/Ouroboros/Network/Protocol/VersionNegotiation/Type.hs

```diff
@@ -0,0 +1,424 @@
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE StandaloneDeriving    #-}
+
+module Ouroboros.Network.Protocol.VersionNegotiation.Type
+  (
+  -- * Protocol Version Info
+    ProtocolVersion (..)
+  , VersionData (..)
+  , protocolVersion
+  , encodeVersionData
+  , decodeVersionData
+  , KnownVersions
+  , newKnownVersions
+  , runKnownVersions
+  , encodeKnownVersions
+  , decodeKnownVersions
+  , findProtocolVersion
+
+  , VersionNegotiationError (..)
+  , ProtocolVersionDescription (..)
+
+  -- * Version Negotiation Protocol
+  , VersionNegotiationProtocol (..)
+  , Message (..)
+  , ClientHasAgency (..)
+  , ServerHasAgency (..)
+  , NobodyHasAgency (..)
+
+  -- ** Version Negotiation client
+  , versionNegotiationClientPeer
+  -- ** Version Negotiation server
+  , versionNegotiationServerPeer
+
+  -- * Auxiliary functions
+  , protocolVersionEq
+  , protocolVersionOrd
+
+  -- * Testing
+  , pureVersionNegotiation
+  )
+  where
+
+
+import           Data.Typeable (Typeable, cast)
+import           Data.List (intersectBy, maximumBy, sortBy)
+
+import           Codec.CBOR.Encoding (Encoding)
+import qualified Codec.CBOR.Encoding as CBOR
+import           Codec.CBOR.Decoding (Decoder)
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Term     as CBOR
+import qualified Codec.Serialise     as CBOR
+import           Codec.Serialise (Serialise (..))
+
+import           Network.TypedProtocol.Core
+
+import           Ouroboros.Network.Mux.Types
+
+-- |
+-- @'ProtocolVersion'@ contains version number @ver@ and associated data, which
+-- is existencially quantified.  It also contains a continuation which takes
+-- local and remote @verData@ and returns the result @r@. @r@ is likely to be
+-- @IO ()@ or something simiar.
+--
+data ProtocolVersion verNumber r where
+  ProtocolVersion
+    :: forall ver verData r.
+       ver
+    -> VersionData ver verData r
+    -> ProtocolVersion ver r
+
+-- |
+-- The @'Eq'@ instance ignores the continuation, it checks if version numbers
+-- are the same and verData have the same type and are equal.
+--
+instance Eq verNumber => Eq (ProtocolVersion verNumber r) where
+  ProtocolVersion ver (VersionData verData _) == ProtocolVersion ver' (VersionData verData' _) =
+    ver == ver' && Just verData == cast verData'
+
+-- |
+-- Data associated with a version.  Its type is existenially quantified in
+-- @'ProtocolVersion'@.  We put @Show@, @Serialise@ and @Typeable@ constraints
+-- needed for running the @'versionNegotiationClient'@ and
+-- @'versionNegotiationServer'@.
+--
+-- It is assumed that @ver@ implies the type of the existencially quantified
+-- @verData@ parameter, otherwise it is a protocol error
+-- @'VersionNegotiationTypeCastError'@.  @Typeable@ constraint is used for safe
+-- casting.
+--
+data VersionData ver verData r where
+  VersionData
+    :: ( Show verData
+       , Eq verData
+       , Serialise verData
+       , Typeable verData
+       )
```

Comment:
I'd suggest not having these here, since `verData` is not existential, so we can put these constraints on as needed in definitions which need them.

Reply:
I'm also quite strongly opposed to using `Serialise` by default. Taking the encoder/decoder itself is more flexible, and an API that uses `Serialise` can easily be layered on top as a derived thing for those who prefer it.

Reply:
Ok I see why these are here: the `verData` is ultimately existential inside the `ProtocolVersion`, so you have no way of getting the constraints you need up there.

Putting these constraints here is IMO not a great idea because it makes it more difficult to use. If I don't need to serialise my version data, why should I have to give a `Serialise` constraint? If I don't need to show it, why do I have to give a `Show` constraint?

I deal with this in my approach by having another type parameter `extra :: Type -> Type` that is applied to the version data type, but appears in the `ProtocolVersion` type. This can be used to plug constraints as needed. You require `Show`? Demand `forall vData . extra vData -> String`. Now the obligation is at the *use* site rather than at construction: I don't need to give it unless it's actually required.